src/uniswap/liquidity.ts

```diff
@@ -453,6 +447,94 @@ export class PoolBalancesAsJson {
 }
 
 export class BalancesStore {
+  constructor(
+    public pairs: {
+      [pair: string]: PairBalances;
+    }
+  ) {}
+
+  public static async load(path: string): Promise<BalancesStore> {
+    try {
+      await access(path);
+    } catch {
+      // If storage file does not exist, which is the default check that `access()` performs, we
+      // just construct an empty store.
+      return new BalancesStore({});
+    }
+
+    try {
+      const data = JSON.parse(await readFile(path, { encoding: "utf8" }));
+
+      if (data === null || typeof data != "object") {
+        throw new Error(`Top level value is not an object in: ${path}`);
+      }
+
+      const pairs: {
+        [pair: string]: PairBalances;
+      } = {};
+      for (const [pair, pairData] of Object.entries(data)) {
+        try {
+          pairs[pair] = await PairBalances.parse(pairData);
+        } catch (err) {
+          throw new Error(`Failed to read balances for: ${pair}\n` + err);
```

Comment:
> It's a little beat strange to have both template literals and strings concatenation at the same time.

The idea is to use concatenation to separate strings that span multiple lines.
This way they also span multiple lines in the source code.
The formatter puts them on the same line here.
But still, it seems easier to see that "\n" when it ends a backtick block, I think.

> More convenient would be to have it like
> ```typescript
> throw new Error(`Failed to read balances for: ${pair}\n${err}`);
> ```

Not sure why exactly it is more convenient.
I would say that it is actually more syntactic noise.
Depends on personal habits, as you may not notice those `$` and `{}`.
But, I think,

```typescript
          throw new Error(`Failed to read balances for: ${pair}\n` + err);
```

is easier to read.

> And of course the best way to write errors is with `cause`

It would be nice to pass in `err` via the `cause` mechanism.
What do you think is the best way to use it?
Add type exceptions on every use?

Reply:
TS 4.6 is going to be released in two weeks and I have tried types the nightly build - they are working as expected. Let's leave it as-is and then fix it after the release